### ISSUE TITLE
Error parsing PEM certs in mbedtls_x509_crt_parse.c

### ISSUE THREAD

**Issue by axelriet** [Mbed-TLS/mbedtls#8555]
### Summary
at mbedtls_x509_crt_parse.c(1412) the statement if (buflen != 0 && buf[buflen - 1] == '\0' && ...) assumes that buf is a string and that buflen includes the null terminator. That's too many assumptions. If the function is called with (chain, string, strlen(string)) where string contains a valid certificate ending in "-----END CERTIFICATE-----" the char tested by buf[buflen - 1] is a '-' and the function decides it's a DER format instead, which subsequently fails to decode at line 1418.

The block 

    if (buflen != 0 && buf[buflen - 1] == '\0' &&
        strstr((const char *) buf, "-----BEGIN CERTIFICATE-----") != NULL) {
        buf_format = MBEDTLS_X509_FORMAT_PEM;
    }

Can be replaced by

   if (buflen != 0 && _strnstr((const char*)buf, "-----BEGIN CERTIFICATE-----",
           min(buflen, 4096)) != NULL) {
       buf_format = MBEDTLS_X509_FORMAT_PEM;
   }

with the three following changes:

1) the zero probing is removed
2) strstr() is replaced by a bounded _strnstr()
3) The search is confined to the first 4096 char, presumably enough to get to the header although RFC 7468 5.2. Explanatory Text does not place a limit, so the value is debatable.

An _strnstr() function should be added somewhere, and the prefix underscore is intended to avoid conflicts if an strnstr() is implemented by some CRT somewhere:

/*
 * Bounded strstr() for use when parsing PEM strings.
 */
char* _strnstr(const char* s, const char* find, size_t slen);

Another, fancier option that does not suffer from the arbitrary 4K preamble limitation could be to implement some memfind() function, e.g. as described here: https://imatix-legacy.github.io/sfl/sfl96.htm#memfind_r

**Comment by davidhorstmann-arm** [comment]
Hi, thanks for raising this!

I think this may be a duplicate of #3896? We currently have a PR #5411 that aims to fix this, but it seems to be stalled.

**Comment by axelriet** [comment]
Yes it looks like both refer to the same thing. I’ll have a look at the PR. I think it’s possible to fix the issue without breaking existing client code. The internal challenge is to find the header without assuming the buffer contains a string, so strstr() cannot be used. Today people pass strlen()+1 to pass that null byte check but this is unnatural and should not be necessary, in particular in APIs that require a byte array and length as input and not just a string :)

**Comment by kleuter** [comment]
What about correctly handling (skipping) comments? For instance, [cacert.pem](https://curl.se/ca/cacert.pem) that contains all Mozilla's certificates cannot be parsed with mbedtls_x509_crt_parse (it thinks the file is in DER format)

**Comment by gilles-peskine-arm** [comment]
@kleuter Parsing comments would be a different feature, so please file a separate issue for that. Unlike the handling of trailing null bytes, that feature would not raise any backward compatibility concern.

**Comment by kleuter** [comment]
@gilles-peskine-arm https://github.com/Mbed-TLS/mbedtls/issues/8812
